src/gleam/http.gleam

```diff
@@ -473,3 +477,98 @@ pub fn set_path(req: Request(body), path: String) -> Request(body) {
     query: query,
   )
 }
+
+fn check_token(token: BitString) {
+  case token {
+    <<"":utf8>> -> Ok(Nil)
+    <<"\s":utf8, _>> -> Error(Nil)
+    <<"\t":utf8, _>> -> Error(Nil)
+    <<"\r":utf8, _>> -> Error(Nil)
+    <<"\n":utf8, _>> -> Error(Nil)
+    <<"\c":utf8, _>> -> Error(Nil)
+    <<"\f":utf8, _>> -> Error(Nil)
+    <<_, rest:bit_string>> -> check_token(rest)
+  }
+}
+
+fn parse_cookie_list(cookie_string) {
+  assert Ok(re) = regex.from_string("[,;]")
+  regex.split(re, cookie_string)
+  |> list.filter_map(
+    fn(pair) {
+      case string.split_once(string.trim(pair), "=") {
+        Ok(tuple("", _)) -> Error(Nil)
+        Ok(tuple(key, value)) -> {
+          try _ = check_token(bit_string.from_string(key))
+          try _ = check_token(bit_string.from_string(value))
+          Ok(tuple(key, value))
+        }
+        Error(Nil) -> Error(Nil)
+      }
+    },
+  )
+}
+
+/// Fetch the cookies sent in a request.
+pub fn get_req_cookies(req) -> List(tuple(String, String)) {
+  let Request(headers: headers, ..) = req
+
+  headers
+  |> list.filter_map(
+    fn(header) {
+      let tuple(key, value) = header
+      case key {
+        "cookie" -> Ok(parse_cookie_list(value))
+        _ -> Error(Nil)
+      }
+    },
+  )
+  |> list.flatten()
+}
+
+pub fn set_req_cookie(req, key, value) {
+  let Request(
+    method: method,
+    headers: headers,
+    body: body,
+    scheme: scheme,
+    host: host,
+    port: port,
+    path: path,
+    query: query,
+  ) = req
+  let new_cookie_string = string.join([key, value], "=")
+
+  let tuple(cookies_string, headers) = case list.key_pop(headers, "cookie") {
+    Ok(tuple(cookies_string, headers)) -> {
+      let cookies_string = string.join(
+        [cookies_string, new_cookie_string],
+        "; ",
+      )
+      tuple(cookies_string, headers)
+    }
+    Error(Nil) -> tuple(new_cookie_string, headers)
+  }
+
+  Request(
+    method: method,
+    headers: [tuple("cookie", cookies_string), ..headers],
+    body: body,
+    scheme: scheme,
+    host: host,
+    port: port,
+    path: path,
+    query: query,
+  )
+}
+
+pub fn set_resp_cookie(resp, key, value, attributes) {
+    let new_cookie_string = string.join([key, value], "=")
+    let attributes = list.map(attributes, fn(attribute) {
```

Comment:
@lpil do have any nice pattern for dealing with lists of options? here attribues is a list of `set_cookie.Option`.

I need to check that an attribute isn't provided twice, and set defaults for certain values. Plug sets a default value for HttpOnly and path (see https://github.com/elixir-plug/plug/blob/master/test/plug/conn/cookies_test.exs#L37)

simply mapping through the list of attributes doesn't give me a clever way to check for duplicates of set a default.
I was thinking of turning it into a dynamic structure and turning the keyword list into a map, but that is relying on me unserstanding the implementation of Gleam types, probably not something to rely on.

Reply:
I'm thinking of having a single Attributes record that you would use as follows
```rust
set_cookie("foo", "bar", Attributes(path: "/", domain: None, HttpOnly: False)
```

I think most people will be working at an API higer than this one so being a bit verbose here doesn't matter